### PR TITLE
Merge ChargeTurret into Turret

### PR DESCRIPTION
**Pull request by MEEPofFaith**
Give any turret the ability to charge. Want a charge burst turret? Now you can. Charge LaserTurret? Yes.
[Video Demonstration](https://youtu.be/c8WeakO-kkM)
(Changes made to blocks shown in this video not included in this pull request.)

### CODE REVIEW

**Comment by SirMachinegun** [comment]
> No no no no non onoonon
> 
> 1. Not really necessary; if you are _that_ desperate for charge liquid/burst turrets, use scripting.
> 2. You will break mods. A lot of them.

alot of mods will be already broken in v6

**Comment by SirMachinegun** [comment]
> You will break mods. A lot of them.

Most mods will be broken in v6 already (mainly because of the new unit production system).I don't think breaking more mods in v6 will be a problem because alot of them are already broken. Also issues because of this can be fixed easily.

File: core/src/mindustry/world/blocks/defense/turrets/Turret.java [hunk 1]
```diff
@@ -59,6 +59,14 @@
     public boolean alternate = false;
     public boolean targetAir = true;
     public boolean targetGround = true;
+    /** Charging related stuff*/
```
Comment (Gdeft) [inline core/src/mindustry/world/blocks/defense/turrets/Turret.java:62]: This won't end well on the java documentation. It'll show like this:

`float chargeTime Charging related stuff`

I assume you are confused by the "Currently used for artillery only" which is for the `minRange` field. You can see it [here](https://mindustrygame.github.io/docs/mindustry/world/blocks/defense/turrets/Turret.html#minRange).

**Review by Gdeft — CHANGES_REQUESTED** [review]

**Comment by sk7725** [comment]
No no no no non onoonon
1. Not really necessary; if you are *that* desperate for charge liquid/burst turrets, use scripting.
2. You will break mods. A lot of them.

**Comment by MEEPofFaith** [comment]
It's not a big break. Just change every ChargeTurret to PowerTurret. And not everyone knows how to script.

File: core/src/mindustry/world/blocks/defense/turrets/Turret.java [hunk 2]
```diff
@@ -304,7 +313,11 @@ protected void turnToTarget(float targetRot){
         }
 
         public boolean shouldTurn(){
-            return true;
+            if(chargeTurn){
+                return true;
+            }else{
+                return !charging;
+            }
```
Comment (summetdev) [inline core/src/mindustry/world/blocks/defense/turrets/Turret.java:320]: I don't know what Anuke prefer, but with ternary operator it looks cleaner.

```suggestion
            return chargeTurn ? true : !charging;
```

File: core/src/mindustry/world/blocks/defense/turrets/Turret.java [hunk 3]
```diff
@@ -343,45 +356,65 @@ protected void updateShooting(){
...
-                    tr.trns(rotation, size * tilesize / 2f, Mathf.range(xRand));
+                    heat = 1f;
+                    recoil = recoilAmount;
+                    
+                    //otherwise, use the normal shot pattern(s)
 
-                    for(int i = 0; i < shots; i++){
-                        bullet(type, rotation + Mathf.range(inaccuracy + type.inaccuracy) + (i - (int)(shots / 2f)) * spread);
+                    if(alternate){
+                        float i = (shotCounter % shots) - (shots-1)/2f;
```
Comment (summetdev) [inline core/src/mindustry/world/blocks/defense/turrets/Turret.java:400]:
```suggestion
                        float i = (shotCounter % shots) - (shots - 1) / 2f;
```

**Comment by sk7725** [comment]
> It's not a big break. Just change every ChargeTurret to PowerTurret. And not everyone knows how to script.

How can you be so sure?
"And not everyone knows how to script." this argument is invalid, *not everyone knows how to mod* either.

**Comment by Anuken** [comment]
> You will break mods. A lot of them.

Every single non-trivial mod is already broken. I've already merged many of the turret types into one; legacy support is not a good reason to keep things as they are.

**Comment by MEEPofFaith** [comment]
So is that a yes, no, vertical, or horizontal?

**Comment by GlennFolker** [comment]
diagonal

**Comment by MEEPofFaith** [comment]
Anuke did it himself, closing pull request.

**Pull request closed without merging**